### ISSUE TITLE
StdoutProxy missing fileno attribute

### ISSUE THREAD

**Issue by vEpiphyte** [prompt-toolkit/python-prompt-toolkit#873]
The StdoutProxy object does not have a fileno attribute.  If the patch_stdout context manager is being held, anything else which may attempt direct access to the stdout fd could fail.  This can be seen by attempting to call faulthandler.dump_traceback() while the context manager is being held.

Example code:

```
import signal
import asyncio
import traceback
import faulthandler

from prompt_toolkit import PromptSession
from prompt_toolkit.patch_stdout import patch_stdout
from prompt_toolkit.eventloop.defaults import use_asyncio_event_loop

def threadstacks(*args, **kwargs):
    '''
    A signal handler used to print thread stacks.
    '''
    print(80 * '*')
    print('Faulthandler stack frames per thread:')
    faulthandler.dump_traceback()
    print(80 * '*')

signal.signal(signal.SIGUSR1, threadstacks)

class Cli:
    def __init__(self):
        # Tell prompt_toolkit to use the asyncio event loop.
        use_asyncio_event_loop()

        self.sess = None
        self.ptext = '> '

    async def prompt(self, text=None):
        if self.sess is None:
            self.sess = PromptSession()
        with patch_stdout():
            retn = await self.sess.prompt(text,
                                          async_=True,
                                          vi_mode=False,
                                          enable_open_in_editor=True)
            return retn

    async def runCmdLoop(self):
        exitloop = False
        while True:
            if exitloop:
                print('Exiting loop...')
                break

            try:

                line = await self.prompt(self.ptext)
                if not line:
                    continue

                line = line.strip()
                if not line:
                    continue

                print(f'Got line: {line}')

                if line in ('quit', 'exit'):
                    exitloop = True

            except KeyboardInterrupt:

                print('<ctrl-c>')
                exitloop = True

            except Exception:
                print('Unknown error occurred:')
                s = traceback.format_exc()
                print(s)
                exitloop = True

async def main():
    cli = Cli()
    await cli.runCmdLoop()

if __name__ == '__main__':
    asyncio.run(main())
```

With that code running, sending SIGUSR1 to the process will terminate it due to an AttributeError

```
$ ps aux | grep break | grep -v grep
newp  5329  0.8  0.1 1191476 24524 pts/12  S+   21:18   0:00 python break_prompt_toolkit.py
$ kill -10 5329
```

With the following error:

```
$ python break_prompt_toolkit.py 
> Unknown error occurred:                                                                                                                                                                                                               
Traceback (most recent call last):
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 560, in run_until_complete
    self.run_forever()
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 528, in run_forever
    self._run_once()
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 1728, in _run_once
    event_list = self._selector.select(timeout)
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/selectors.py", line 468, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
  File "break_prompt_toolkit.py", line 16, in threadstacks
    faulthandler.dump_traceback()
AttributeError: 'StdoutProxy' object has no attribute 'fileno'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "break_prompt_toolkit.py", line 48, in runCmdLoop
    line = await self.prompt(self.ptext)
  File "break_prompt_toolkit.py", line 36, in prompt
    enable_open_in_editor=True)
concurrent.futures._base.CancelledError

Exiting loop...
********************************************************************************
> Traceback (most recent call last):                                                                                                                                                                                                    
  File "break_prompt_toolkit.py", line 77, in <module>
    asyncio.run(main())
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 560, in run_until_complete
    self.run_forever()
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 528, in run_forever
    self._run_once()
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 1728, in _run_once
    event_list = self._selector.select(timeout)
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/selectors.py", line 468, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
  File "break_prompt_toolkit.py", line 16, in threadstacks
    faulthandler.dump_traceback()
AttributeError: 'StdoutProxy' object has no attribute 'fileno'
```

Environment:
Ubuntu 18.04; Python 3.7.1 installed via pyenv

Prompt-toolkit version
```
prompt-toolkit==2.0.9
```

**Comment by jonathanslenders** [comment]
Hi @vEpiphyte,
Thank you for pointing out this issue!
This should probably fix it: https://github.com/prompt-toolkit/python-prompt-toolkit/pull/874
Can you try again, using the master branch?

**Comment by vEpiphyte** [comment]
@jonathanslenders I just tested my script against the master branch and it behaved probably as well as we can expect.  There is some weird ordering due to print() calls occuring before and after faulthandler but this is better than dying :)

Thank you for the quick turnaround!

```> help                                                                                                                                           
Got line: help
> wordsCurrent thread 0x00007fd085157080 (most recent call first):                                                                               
  File "break_prompt_toolkit.py", line 16 in threadstacks
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/selectors.py", line 468 in select
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 1728 in _run_once
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 528 in run_forever
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/base_events.py", line 560 in run_until_complete
  File "/home/newp/.pyenv/versions/3.7.1/lib/python3.7/asyncio/runners.py", line 43 in run
  File "break_prompt_toolkit.py", line 77 in <module>
********************************************************************************
Faulthandler stack frames per thread:
********************************************************************************
```

**Comment by vEpiphyte** [comment]
Resolved in #874
